plan/expr_to_pb.go

```diff
@@ -0,0 +1,331 @@
+
+package plan
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/types"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+func expressionsToPB(exprs []expression.Expression, client kv.Client) (pbExpr *tipb.Expr, remained []expression.Expression, err error) {
+	for _, expr := range exprs {
+		v, err := exprToPB(client, expr)
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+		if v == nil {
+			remained = append(remained, expr)
+			continue
+		}
+		if pbExpr == nil {
+			pbExpr = v
+		} else {
+			// merge multiple converted pb expression into an AND expression.
+			pbExpr = &tipb.Expr{
+				Tp:       tipb.ExprType_And,
+				Children: []*tipb.Expr{pbExpr, v}}
+		}
+	}
+	return
+}
+
+func exprToPB(client kv.Client, expr expression.Expression) (*tipb.Expr, error) {
+	switch x := expr.(type) {
+	case *expression.Constant:
+		return datumToPBExpr(client, expr.(*expression.Constant).Value)
+	case *expression.Column:
+		return columnToPBExpr(client, x)
+	case *expression.ScalarFunction:
+		return scalarFuncToPBExpr(client, x)
+	}
+	return nil, nil
+}
+
+func datumToPBExpr(client kv.Client, d types.Datum) (*tipb.Expr, error) {
+	var tp tipb.ExprType
+	var val []byte
+	switch d.Kind() {
+	case types.KindNull:
+		tp = tipb.ExprType_Null
+	case types.KindInt64:
+		tp = tipb.ExprType_Int64
+		val = codec.EncodeInt(nil, d.GetInt64())
+	case types.KindUint64:
+		tp = tipb.ExprType_Uint64
+		val = codec.EncodeUint(nil, d.GetUint64())
+	case types.KindString:
+		tp = tipb.ExprType_String
+		val = d.GetBytes()
+	case types.KindBytes:
+		tp = tipb.ExprType_Bytes
+		val = d.GetBytes()
+	case types.KindFloat32:
+		tp = tipb.ExprType_Float32
+		val = codec.EncodeFloat(nil, d.GetFloat64())
+	case types.KindFloat64:
+		tp = tipb.ExprType_Float64
+		val = codec.EncodeFloat(nil, d.GetFloat64())
+	case types.KindMysqlDuration:
+		tp = tipb.ExprType_MysqlDuration
+		val = codec.EncodeInt(nil, int64(d.GetMysqlDuration().Duration))
+	case types.KindMysqlDecimal:
+		tp = tipb.ExprType_MysqlDecimal
+		val = codec.EncodeDecimal(nil, d)
+	default:
+		return nil, nil
+	}
+	if !client.SupportRequestType(kv.ReqTypeSelect, int64(tp)) {
+		return nil, nil
+	}
+	return &tipb.Expr{Tp: tp, Val: val}, nil
+}
+
+func columnToPBExpr(client kv.Client, column *expression.Column) (*tipb.Expr, error) {
+	if !client.SupportRequestType(kv.ReqTypeSelect, int64(tipb.ExprType_ColumnRef)) {
+		return nil, nil
+	}
+	switch column.GetType().Tp {
+	case mysql.TypeBit, mysql.TypeSet, mysql.TypeEnum, mysql.TypeGeometry, mysql.TypeDecimal:
+		return nil, nil
+	}
+
+	if column.Correlated {
+		return nil, nil
+	}
+
+	id := column.ID
+	// Zero Column ID is not a column from table, can not support for now.
+	if id == 0 {
+		return nil, nil
+	}
+	// its value is available to use.
```

Comment:
It is ported from executor written by @zimulala . I guess she ported this from old plan. So maybe no one know why.